### PR TITLE
#3 submission of registration form

### LINKED ISSUE #3: Submission of Registration Form
**Trigger:**

The student clicked the submit button

**Page:**

Image

**Requirements:**

- [x] The endpoint should be '/student/submit-registration'
- [x] Should automatically create a UUID
- [x] The UUID, name, email, year_and_course, and payment_photo_link (payment.photo_src) should be saved on the supabase database
- [x] The payment photo should be stored on the supabase storage
- [x] The status should be pending if it was registered by the student but if it was registered by the cashier, it would be accepted.

### PR DESCRIPTION
**Pull request by Miraii133**
chipi chipi chapa chapa dubi dubi daba daba

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -26,7 +26,7 @@ model Student {
   paymentId Int     @unique
 
   event   Event @relation(fields: [eventId], references: [id])
-  eventId Int
+  eventId Int   @unique
```
Comment (rhaeyx) [inline prisma/schema.prisma]: baii, giremove ang `@unique` diri para maka many (students) to one (event) na relationship

File: prisma/migrations/20231229154115_update_to_new_schema/migration.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+-- AlterTable
```
Comment (rhaeyx) [inline prisma/migrations/20231229154115_update_to_new_schema/migration.sql]: pwede patanggal neto sa next commit, meron na migration neto sa main branch named added_requires_payment. mag error man gud ang prisma:setup if duplicate migrations

File: src/student/dto/create-student.dto.ts [hunk 3]
```diff
@@ -1,4 +1,11 @@
 export class CreateStudentDto {
+  createdAt?: Date | string;
+  updatedAt?: Date | string;
+  uuid: string;
   firstName: string;
   lastName: string;
+  email: string;
+  year_and_course: string;
+  paymentId?: number | null;
```
Comment (rhaeyx) [inline src/student/dto/create-student.dto.ts]: di naga throw ng error sayo yung `paymentId?` ?

```
[7:49:42 PM] Starting compilation in watch mode...

src/student/student.service.ts:64:7 - error TS2322: Type 'CreateStudentDto' is not assignable to type '(Without<StudentCreateInput, StudentUncheckedCreateInput> & StudentUncheckedCreateInput) | (Without<...> & StudentCreateInput)'.
  Type 'CreateStudentDto' is not assignable to type 'Without<StudentUncheckedCreateInput, StudentCreateInput> & StudentCreateInput'.
    Type 'CreateStudentDto' is not assignable to type 'Without<StudentUncheckedCreateInput, StudentCreateInput>'.
      Types of property 'paymentId' are incompatible.
        Type 'number' is not assignable to type 'never'.

64       data: createStudentDto,
         ~~~~

  node_modules/.prisma/client/index.d.ts:2113:5
    2113     data: XOR<StudentCreateInput, StudentUncheckedCreateInput>
             ~~~~
    The expected type comes from property 'data' which is declared here on type '{ select?: StudentSelect<DefaultArgs>; include?: StudentInclude<DefaultArgs>; data: (Without<StudentCreateInput, StudentUncheckedCreateInput> & StudentUncheckedCreateInput) | (Without<...> & StudentCreateInput); }'

[7:49:46 PM] Found 1 error. Watching for file changes.
```

mag ganto sakin after npm run start:dev if may question mark dito. also, sabay man pagkasubmit and create ang payment and student so pwede lang wala yang `?`

File: src/student/student.service.ts [hunk 4]
```diff
@@ -3,21 +3,96 @@ import { CreateStudentDto } from './dto/create-student.dto';
...
+
+  async createStudent(
+    createStudentDto: CreateStudentDto,
+    file: Express.Multer.File,
+  ) {
+    const uuid = uuidv4();
+    const payment_path = this.uploadImageToDB(file, uuid);
+    createStudentDto.uuid = uuid;
+    const isStudent = true;
+    const payment = await this.createPayment(await payment_path, isStudent);
```
Comment (rhaeyx) [inline src/student/student.service.ts]:
```
    const payment_path = this.uploadImageToDB(file, uuid);
    createStudentDto.uuid = uuid;
    const isStudent = true;
    const payment = await this.createPayment(await payment_path, isStudent);
```
any reason bakit nasa `payment_path` ang await and wala sa `const payment_path = await this.uploadImageToDB(file, uuid)`? if okay lang pwede palipat sa this.upload yung await

File: src/student/student.service.ts [hunk 5]
```diff
@@ -3,21 +3,96 @@ import { CreateStudentDto } from './dto/create-student.dto';
...
+  }
+
+  async createPayment(payment_path: string, isRegisterByStudent: boolean) {
+    const newPayment = await this.prisma.payment.create({
+      data: {
+        photo_src: payment_path,
+        // if submission is by student, put in pending status
+        // until cashier accepts payment
+        // else submission is by cashier, accepted status
+        status: isRegisterByStudent ? 'pending' : 'accepted',
```
Comment (rhaeyx) [inline src/student/student.service.ts]: okay na eto for now, but once ma setup na ang authentication, check should be done with the session variables. kung ano user type ng nag submit nung request.

File: src/student/student.service.ts [hunk 6]
```diff
@@ -3,21 +3,96 @@ import { CreateStudentDto } from './dto/create-student.dto';
...
+      },
+    });
+    return newPayment;
+  }
+
+  async createEvent() {
```
Comment (rhaeyx) [inline src/student/student.service.ts]: paapil ko remove ani na function men

File: src/student/student.service.ts [hunk 7]
```diff
@@ -3,21 +3,96 @@ import { CreateStudentDto } from './dto/create-student.dto';
...
+    return newEvent;
+  }
+
+  async createStudent(
+    createStudentDto: CreateStudentDto,
+    file: Express.Multer.File,
+  ) {
+    const uuid = uuidv4();
+    const payment_path = this.uploadImageToDB(file, uuid);
+    createStudentDto.uuid = uuid;
```
Comment (rhaeyx) [inline src/student/student.service.ts]: `createStudentDto.uuid = uuidv4()` nalang siguro dretso para wala na ng `const uuid = uuidv4()`

File: src/student/student.service.ts [hunk 8]
```diff
@@ -3,21 +3,96 @@ import { CreateStudentDto } from './dto/create-student.dto';
...
     });
+    return newStudent;
+  }
 
-    return testStudent;
+  toBase64(file: Express.Multer.File) {
+    return Buffer.from(file.buffer).toString('base64');
+  }
+
+  async uploadImageToDB(file: Express.Multer.File, uuid: string) {
```
Comment (rhaeyx) [inline src/student/student.service.ts]: this function should be inside supabaseservice also pa edit ng .env.example sa mga ginamit mo na env variables

File: src/student/student.service.ts [hunk 9]
```diff
@@ -3,21 +3,96 @@ import { CreateStudentDto } from './dto/create-student.dto';
...
+import { decode } from 'base64-arraybuffer';
+import { v4 as uuidv4 } from 'uuid';
 
 @Injectable()
 export class StudentService {
   constructor(private prisma: PrismaService) {}
 
-  create(createStudentDto: CreateStudentDto) {
-    const testStudent = this.prisma.testModel.create({
+  getPaymentById(createStudentDto) {
```
Comment (rhaeyx) [inline src/student/student.service.ts]: not sure if this function is used somewhere, if wala pa remove nalang din

File: src/student/student.service.ts (hunk 6 above)
Comment (rhaeyx) [inline src/student/student.service.ts]: pwede na eto iremove na function, instead need nalang iconnect sa already existing event

File: src/student/dto/create-student.dto.ts
Comment (rhaeyx) [inline src/student/dto/create-student.dto.ts:1]: Please add validation nalang din sa lahat ng fields and check the mimetype ng uploaded file (only allow png, jpeg, jpg, heic, and other common image file types). btw meron included na validation pipes ang nest
https://docs.nestjs.com/techniques/validation

p.s: also try catches din

**Review by rhaeyx — CHANGES_REQUESTED** [review]

File: prisma/migrations/20231229154115_update_to_new_schema/migration.sql (hunk 2 above)
Comment (Miraii133) [reply prisma/migrations/20231229154115_update_to_new_schema/migration.sql]: Resolved, thanks!
Action:
-Remove prisma/migrations/20231229154115_update_to_new_schema/migration.sql file

File: prisma/schema.prisma (hunk 1 above)
Comment (Miraii133) [reply prisma/schema.prisma]: Resolved, thanks!
Action:
-Remove @unique in eventId

File: src/student/student.service.ts (hunk 4 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Resolved, thanks!
Action:
-Move await to this.uploadImageToDB()

File: src/student/student.service.ts (hunk 5 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Question:
Pag naa nay OAuth, mag assume ko naa na siyay session id or unsa ba na identifier sa user type, ma send na siya sa Body sa POST method? @rhaeyx

Comment (rhaeyx) [reply src/student/student.service.ts]: Yeah, pag ma setup na ang auth kay part sa session ang user type. Gi note lang nako daan basig makalimot ko ug remind pag ma implement na ang auth. Di pa ko sure asa mabutang ang data though, depende pa sa mag implement.

File: src/student/student.service.ts (hunk 9 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Resolved, thanks!
Action:
-Remove getPaymentById()

File: src/student/student.service.ts (hunk 6 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Question:
Kung need i connect sa already existing event, dapat mag expect ko sa /submit-registration na endpoint kay naay eventId noh? Then didto na nako i connect? @rhaeyx

Comment (rhaeyx) [reply src/student/student.service.ts]: Yes, supplied ang eventId sa form then connect nalang nimo.

File: src/student/student.service.ts (hunk 8 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Questions:
1. Asa na folder nako i create ang supabase service na file? Sa may supabase na folder? or sa student lang pud?
2. Pwede ko pa explain further unsa pasabot nimo sa pa edit sa .env? I change nako ang `process.env.SUPABASE_SERVICE_ROLE_KEY, process.env.SUPABASE_URL`? 
@rhaeyx

File: src/student/dto/create-student.dto.ts (hunk 3 above)
Comment (Miraii133) [reply src/student/dto/create-student.dto.ts]: Dili siya naga error saakoa pag himuong optional ang paymentId. Nag mention kasi si @viverbungag na naay instances na dili na need mag pay si Student so pwede lang siya i null.

File: src/student/student.service.ts (hunk 5 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Ahh noted noted. Pwede nani i mark as resolve or i wait sa nako ang pag implement sa OAuth para dili makalimtan?

Comment (rhaeyx) [reply src/student/student.service.ts]: Iresolve lang, tas pa remind nalang pud or create ra ug another PR pag ma implement na ang auth.

File: src/student/student.service.ts (hunk 7 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Gi buhatan nako ang uuid ug own variable niya kay para ma pass siya sa `uploadImageToDB()` kay basig pag mag `this.uploadImageToDB(file, uuidv4());` ko kay basig mag create siyag lahi na uuid na dili parehas sa `createStudentDto.uuid` 

Tama ba?

File: src/student/dto/create-student.dto.ts (hunk 3 above)
Comment (rhaeyx) [reply src/student/dto/create-student.dto.ts]: Ahhhh bitaw tama tama, pwede man diay wala'y payment. Pa keep nalang as is Jiyo, lamatss!

File: src/student/student.service.ts (hunk 7 above)
Comment (rhaeyx) [reply src/student/student.service.ts]: Ay wa nako na notice na ma pass siya sa uploadImageToDb. Tama na, mbmb thank u!

File: src/student/student.service.ts (hunk 5 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Resolved, thanks!
Action:
-Reminder to implement user type once OAuth is created

File: src/student/student.service.ts (hunk 7 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Resolved, thanks!
Action:
-None

File: src/student/dto/create-student.dto.ts (hunk 3 above)
Comment (Miraii133) [reply src/student/dto/create-student.dto.ts]: Resolved, thanks!
Action:
-None

File: src/student/student.service.ts (hunk 8 above)
Comment (rhaeyx) [reply src/student/student.service.ts]: 1. Sa 1-login-xxx na branch kay nag gamit siya supabaseService didto, pwede ka mag refer ato pero modify lang as you see fit.

2. Pa set lang ko ug placeholders sa .env.example na file, like add tung mga variables na gidungag nimo sa .env
```
...
DATABASE_URL="postgresql://postgres:postgres@localhost:54322/postgres?schema=public"
DIRECT_URL="postgresql://postgres:postgres@localhost:54322/postgres?schema=public"
SUPABASE_URL="http://127.0.0.1:54321"
SUPABASE_ANON_KEY="supabase_anon_key"
SUPABASE_SERVICE_ROLE_KEY="supabase_service_role_key"
```
so ma something like this na ang .env.example

Comment (Miraii133) [reply src/student/student.service.ts]: Resolved, thanks!
Action:
-Moved uploadImageToDB() and toBase64() functions to supabase.service
-Copied and created supabase.service and supabase.module from issue 1 (might merge conflict)
-Added new placeholders to .env.example file

File: src/student/dto/create-student.dto.ts
Comment (Miraii133) [reply src/student/dto/create-student.dto.ts:1]: Resolved, thanks!
Actions:
-Added new function to validate image file types.
-Added new try catch to functions using prisma.

File: .env.example [hunk 10]
```diff
@@ -5,6 +5,10 @@
...
 # Rename this file to .env
-
-DATABASE_URL="postgresql://postgres:postgres@localhost:54322/postgres?schema=public"
-DIRECT_URL="postgresql://postgres:postgres@localhost:54322/postgres?schema=public"
+PORT=5000
+NODE_ENV=development
+DATABASE_URL=postgresql://postgres:postgres@127.0.0.1:54322/postgres
+DIRECT_URL=postgresql://postgres:postgres@127.0.0.1:54322/postgres
+SUPABASE_URL=http://127.0.0.1:54321
+SUPABASE_KEY="supabase_key"
```
Comment (rhaeyx) [inline .env.example:13]: ang service role key pasabot ani no?

File: src/student/student.service.ts [hunk 11]
```diff
@@ -1,23 +1,81 @@
...
+  }
+
+  async createStudent(
+    createStudentDto: CreateStudentDto,
+    file: Express.Multer.File,
+  ) {
+    this.ValidateFileType(file);
+    const uuid = uuidv4();
+    const payment_path = await this.supabaseService.uploadImageToDB(file, uuid);
+    createStudentDto.uuid = uuidv4();
```
Comment (rhaeyx) [inline src/student/student.service.ts]:
```
const uuid = uuidv4();
const payment_path = await this.supabaseService.uploadImageToDB(file, uuid);
createStudentDto.uuid = uuidv4();
```
    
Intended eto na mag gawa ng bagong uuid string and hindi i reuse yung previously generated uuid?

File: src/student/student.controller.ts [hunk 12]
```diff
@@ -6,25 +6,36 @@ import {
...
   @Get()
   findAll() {
+    console.log('find all executed');
     return this.studentService.findAll();
   }
+  // endpoint for when user clicks
+  // submit button
+  // fileIntercepter accepts any files sent
+  // through POST body
+  @Post('submit-registration')
```
Comment (rhaeyx) [inline src/student/student.controller.ts:27]: pa tanggal nalng ko sa comments diri na part broo

File: src/event/event.service.ts [hunk 13]
```diff
@@ -9,6 +9,8 @@ export class EventService {
   async viewEvent(eventId: number) {
     return await this.prismaService.event.findFirst({
       where: { id: eventId },
+      // replaced students with student, students is not
+      // found and causes error
```
Comment (rhaeyx) [inline src/event/event.service.ts]: patanggal nalang din neto na comment

**Review by rhaeyx — CHANGES_REQUESTED** [review]
Minor changes nalang then ready to merge na. Pa chat lang if tapos na!

File: .env.example (hunk 10 above)
Comment (Miraii133) [reply .env.example:13]: yessir actually dili ko sure kung apilon ang anon key

File: src/student/student.controller.ts (hunk 12 above)
Comment (Miraii133) [reply src/student/student.controller.ts:27]: Resolved, thanks!
Action:
-Removed comments

File: src/event/event.service.ts (hunk 13 above)
Comment (Miraii133) [reply src/event/event.service.ts]: Resolved, thanks!
Action:
-Removed comments

File: src/student/student.service.ts (hunk 11 above)
Comment (Miraii133) [reply src/student/student.service.ts]: Resolved, thanks!
Action:
-Changed used uuid variable for createStudentDto.uuid instead of generating another uuid

**Comment by Miraii133** [comment]
Made a mistake on handling branches, reverted to last commit and had to re-implement all of the new PR comments. All goods now.

**Pull request merged**